[7.0.x] custom planet containers

a-palchikov opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
Forward port of https://github.com/gravitational/gravity/pull/1962.
Also fixes an issue with the upgrade unable to determine existing etcd server version if the planet container did not expose this information.

## Type of change
<!--Required. Keep only those that apply.-->

* Bug fix (non-breaking change which fixes an issue)
* Regression fix (non-breaking change which fixes a regression)
* This change has a user-facing impact

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Refs https://github.com/gravitational/gravity/issues/1940.
<!--This PR depends on the following PRs (e.g. planet, satellite, etc.).-->
* Requires https://github.com/gravitational/planet/pull/723, https://github.com/gravitational/gravity.e/pull/4328.
<!--This PR is a back-/forward-port of the following PR.-->
* Ports https://github.com/gravitational/gravity/pull/1962.

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing
- [ ] Address review feedback

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
#### Prerequisites
1. Custom planet container replacing the default package:

```yaml
systemOptions:
  baseImage: quay.io/gravitational/planet:<custom-build>
```

2. Custom planet container for a (subset of) node profile

```yaml
nodeProfiles:
  - name: profile
    systemOptions:
      baseImage: quay.io/gravitational/planet:<custom-build>
```

#### Clusters

- [x] Installation, 3-node cluster, all control plane nodes
- [x] Installation, 3-node cluster, one control plane, 2 workers
- [x] Upgrade from a previous version (`6.1.33`), 3-node cluster
- [ ] Upgrade from a previous version, run gc and expand (both controller plane and workers)
- [ ] Installation, 1-node cluster,  then expand to 3-nodes (both controller plane and workers)

a-palchikov committed:
Forward port https://github.com/gravitational/gravity/pull/1962.

a-palchikov commented on lib/install/phases/postsystem.go:
```diff
@@ -164,7 +164,7 @@ type healthExecutor struct {
 	p.Progress.NextStep("Waiting for the planet to start")
 	p.Info("Waiting for the planet to start.")
-	err := utils.Retry(defaults.RetryInterval, defaults.RetryAttempts,
+	err := utils.RetryWithInterval(ctx, defaults.ExponentialBackOff(),
```
Found this issue that prevents timely abort of the installer during the planet status wait phase.

a-palchikov commented on lib/update/cluster/builder_test.go:
```diff
@@ -391,113 +391,50 @@ func (s *PlanSuite) TestPlanWithIntermediateRuntimeUpdate(c *check.C) {
 	})
 }
 
-func (s *PlanSuite) TestUpdatesEtcdFromManifestWithoutLabels(c *check.C) {
```
Removed the test about assumed etcd version since it is expected that we have both installed and upgrade versions available.

a-palchikov committed:
Renmove unnecessary abstractions like NewSyncerFunc and GetRepositoryFunc
for a clearer comprehension of the builder.
Expose the build API key to be able to use hub/token combo to access
distribution hub regardless of the login context.

a-palchikov committed:
Fix package sync to also fetch the application package

a-palchikov committed:
Use hub/token flags to pull the necessary installers for robotest

a-palchikov committed:
Expose distribution hub key for the robotest job

wadells commented on build.assets/robotest/run.sh:
```diff
@@ -27,6 +27,7 @@ export GCE_VM=${GCE_VM:-custom-4-8192}
 # Parallelism & retry, tuned for GCE
 export PARALLEL_TESTS=${PARALLEL_TESTS:-4}
 export REPEAT_TESTS=${REPEAT_TESTS:-1}
+GRAVITY_PULL_OPTIONS=${GRAVITY_PULL_OPTIONS:-""}
```
This doesn't appear to change anything for CI robotest runs -- I don't see a reference to this variable anywhere else in our Makefiles & toolchain. Is this for local testing?

a-palchikov replied:
It is coming from [outside](https://github.com/gravitational/gravity.e/pull/4328/files#diff-b67911656ef5d18c4ae36cb6741b7965R30)

wadells commented on Jenkinsfile:
```diff
@@ -84,6 +84,7 @@ timestamps {
                 file(credentialsId:'ROBOTEST_LOG_GOOGLE_APPLICATION_CREDENTIALS', variable: 'GOOGLE_APPLICATION_CREDENTIALS'),
                 file(credentialsId:'OPS_SSH_KEY', variable: 'SSH_KEY'),
                 file(credentialsId:'OPS_SSH_PUB', variable: 'SSH_PUB'),
+                string(credentialsId:'GET_GRAVITATIONAL_IO_APIKEY', variable: 'BUILD_APIKEY'),
```
We shouldn't need this to fetch images, as far as I know. `tele pull` will download the correct artifacts when it isn't logged in to get.gravitational.io -- so long as it isn't logged in elsewhere either.

We should test both authenticated & unauthenticated fetches, but if we only do one, unauthenticated are more important to customers in my opinion.  Especially for the new user experience.

a-palchikov replied:
Do we always logout? Since if there is login state (from an unrelated job), this will still use it (this is what's been happening here).

wadells replied:
We did briefly, but that caused issues for publishing.

I think login is the way to go for now -- it is consistent with the old behavior.

I'm working on a patch that containerizes more of the robotest workflow such that `~jenkins` state cannot affect fetching robotest artifacts (and vice versa), but that is probably a day or two out -- it is a more substantial refactoring

a-palchikov replied:
I have nothing against using login but it seems that publishing is a very focused task which is not necessary for PR jobs - I think explicit token configuration is as good as logging in in a per-task environment, no?

wadells replied:
> I have nothing against using login but it seems that publishing is a very focused task which is not necessary for PR jobs

Publishing & PR shouldn't be related, but are due to sharing login state in `~jenkins`.  See https://jenkins.gravitational.io/view/Publishing/job/Telekube-Publish-5.2/301/console for an example failure

> I think explicit token configuration is as good as logging in in a per-task environment, no?

My goal for test purposes would be to fetch artifacts while not logged in.  I've yet to find a good way to use unauthenticated pulls in a way that is non

On a more holistic note, I want to isolate because sharing state in `~/` has proven to be prone to races and a large pain point.  This stolon/pithos CI breaking robotest topic is the 3rd notable disruption stemming from "spooky action at a distance" when unrelated parts of our CI/release infra breaking each other due to sharing state in `~/`.  See https://github.com/gravitational/gravity/pull/1941 and https://github.com/gravitational/gravity/pull/177 for the first two.

While I *think* I understand enough to not make mistakes regarding this shared state after having been burned by it twice, will the next employee who touches our release infra understand?  Why leave this edge around for them to trip on?

a-palchikov replied:
I think we're not following each other. I'm not saying we should use shared state and using hub/token does not create shared (login) state - it is a way to use the login information at that particular point where it is applied - isn't this what we're after?

wadells replied:
> I'm not saying we should use shared state and using hub/token does not create shared (login) state - it is a way to use the login information at that particular point where it is applied - isn't this what we're after?

A couple issues with the hub/token approach:

1) How do we apply this advice to 6.1 and 5.5 where the hub/token flags do not exist?  Those branches are subject to the same shared state modes of failure. 

2) How do we test unauthenticated fetch with the hub and token flags?  Related: How could open source users run any of these targets?

To be clear: I'm not asking you to change anything here in this PR.  We need an immediate mitigation and your PR provides this in the best way!  This discussion is about a long term plan.

wadells replied:
The answer to all the above could be "product fixes & backports to all active releases".  But per our earlier discussion about "do unnecessary feature backports" I'm inclined to look at non-product fixes to our build architecture.  Lets fix the product in master, but use non-product solutions (could be sandboxing pushes/pulls in a container, could be something else) to fix the issue in older branches.  This reduces surface area for product regressions, especially since this wouldn't be work a customer cares about -- it is purely internal work.

bernardjkim commented on lib/builder/builder.go:
```diff
@@ -270,13 +246,9 @@ func (b *Builder) SyncPackageCache(ctx context.Context, runtimeVersion semver.Ve
 	for _, runtimeVersion := range append([]semver.Version{runtimeVersion}, intermediateVersions...) {
 		b.NextStep("Syncing packages for %v", runtimeVersion)
-		if err := b.syncPackageCache(ctx, runtimeVersion, syncer, apps); err != nil {
+		if err := b.syncPackageCache(ctx, runtimeVersion, b.Syncer, apps); err != nil {
```
nit: can get `b.Syncer` from inside `syncPackageCache`

bernardjkim commented on lib/builder/syncer.go:
```diff
@@ -41,122 +39,86 @@ type Syncer interface {
-
 // s3Syncer synchronizes local package cache with S3 bucket
-type s3Syncer struct {
+type S3Syncer struct {
```
nit: Update godoc `s3Syncer` -> `S3Syncer`

a-palchikov committed:
Address review comments

a-palchikov committed:
Use build API key name as used in robotest in other scripts consistently

r0mant reviewed (commented):
I think that this massive builder refactoring makes this a pretty risky change esp. for an LTS release. My worry is we'll break one of our subtle tele build use-cases as we did in the past.

I would either see if it's possible to fix the custom planet problem with less of unnecessary refactoring, or extensively test all tele build use-cases (for both enterprise and open-source where applicable) to make sure nothing regresses:

- [ ] That when logged out, it pulls stuff from default hub.
- [ ] That when logged into a hub (both via "tele login" and "tsh login"), it uses that hub.
- [ ] That when --hub/--token/--repository provided, it uses the indicated hub.
- [ ] Behavior of --state-dir flag, should be source for both packages and credentials (when logged in via tele login --state-dir), and when used in conjunction with --hub/--token/--repository flags.

Maybe I'm forgetting something else...

a-palchikov committed:
Use the dependency iteration APIs to query packages/applications

a-palchikov commented:
Checked the following for enterprise builds:

- [x] That when logged out, it pulls stuff from default hub.
- [x] That when logged into a hub (both via "tele login" and "tsh login"), it uses that hub.
- [x] [!] That when --hub/--token/--repository provided, it uses the indicated hub. **This did not work when using --token/--repository combination - seems like --hub is always required**.
- [x] Behavior of --state-dir flag, should be source for both packages and credentials (when logged in via tele login --state-dir), and when used in conjunction with --hub/--token/--repository flags.

dorsany commented:
Hi @a-palchikov, do you think that this pr will be a part of the next release version of `7.0.x` ?

a-palchikov commented:
@dorsany I have a couple more things to take care as part of this PR that were discovered in the neighbor branch, but I hope it will.

a-palchikov committed:
Update docker client to fix data races found in builder tests

r0mant commented on lib/pack/localpack/packageserver.go:
```diff
@@ -327,22 +327,14 @@ func (p *PackageServer) UpsertRepository(repository string, expires time.Time) e
 	}
-	_, err = p.backend.CreateRepository(repo)
-	if err != nil {
+	_, err := p.backend.CreateRepository(repo)
```
Bit weird that the function is called "UpsertRepository" yet it only creates it. Prob. not a big deal since it's been like this forever, only just noticing.

a-palchikov replied:
This should propagate to the storage interface which should offer `UpsertRepository` in the same vein. Not going to change that though in this PR.

r0mant commented on lib/update/cluster/plan.go:
```diff
@@ -425,26 +425,18 @@ func shouldUpdateCoreDNS(client *kubernetes.Clientset) (bool, error) {
-	installedVersion, err := getEtcdVersion(*runtimePackage, packages)
+	installedVersion, err := getEtcdVersion(installedRuntimePackage, packages)
 	if err != nil {
 		if !trace.IsNotFound(err) {
```
Shouldn't "not found" also be a terminal scenario in this case? Since the etcd migrate operation is now basically copying old directory to new directory, the old one must always be present. Otherwise "installedVersion" below will be empty and we'll run into the same scenario as the customer S reported to us yesterday. Am I missing something?

a-palchikov replied:
It should be the other way around - the etcd data directory migration path needs to support the case of the empty source version and fallback to a version file as it used to be.

r0mant commented on lib/update/cluster/steps.go:
```diff
@@ -71,7 +71,15 @@ func (r *phaseBuilder) initSteps(ctx context.Context) error {
 		return trace.Wrap(err)
 	}
-	etcd, err := shouldUpdateEtcd(installedRuntimeApp, r.updateRuntimeApp, r.packages)
+	installedRuntimePackage, err := schema.GetDefaultRuntimePackage(r.installedApp.Manifest)
```
Does this properly handle custom planets? It's just the name "Get**Default**RuntimePackage" gives the impression that it always uses the default planet.

a-palchikov replied:
Yeah, it is confusing. I think I'll drop the default part.

r0mant reviewed (approved):
Left a couple of question, otherwise lgtm.

a-palchikov committed:
Handle empty source etcd version in migrate path - in this case, read the version file to get this detail.

a-palchikov committed:
Address review comments.
Read the etcd cluster version when creating the upgrade operation to use
as a fallback in case the planet image does not have the etcd version
information in the package manifest as expected.

r0mant commented on lib/update/cluster/plan.go:
```diff
@@ -475,6 +483,35 @@ func getEtcdVersion(locator loc.Locator, packageService pack.PackageService) (*s
+func getCurrentEtcdVersion(stateDir string) (version *semver.Version, err error) {
+	// https://github.com/gravitational/planet/blob/version/7.0.x/tool/planet/constants.go#L87
+	const etcdVersionKey = "PLANET_ETCD_VERSION"
+	f, err := os.Open(filepath.Join(state.EtcdDir(stateDir), "etcd-version.txt"))
```
I'm curious, instead of relying on this file being present, why not just do `etcd --version` (or a respective API call)?

a-palchikov replied:
This an excellent point - I've addressed it in https://github.com/gravitational/gravity/pull/1982/commits/7c19daac69f80530b0eaaf888083aa267ba14593

a-palchikov committed:
Address review comments

a-palchikov committed:
Bump e

a-palchikov committed:
Update e to branch release
